packages/babel-parser/src/parser/expression.js

```diff
@@ -1223,19 +1273,22 @@ export default class ExpressionParser extends LValParser {
     }
   }
 
-  parseFunctionExpression(): N.FunctionExpression | N.MetaProperty {
+  parseFunctionOrFunctionSent(): N.FunctionExpression | N.MetaProperty {
     const node = this.startNode();
 
-    // We do not do parseIdentifier here because when parseFunctionExpression
+    // We do not do parseIdentifier here because when parseFunctionOrFunctionSent
     // is called we already know that the current token is a "name" with the value "function"
     // This will improve perf a tiny little bit as we do not do validation but more importantly
     // here is that parseIdentifier will remove an item from the expression stack
     // if "function" or "class" is parsed as identifier (in objects e.g.), which should not happen here.
-    let meta = this.startNode();
-    this.next();
-    meta = this.createIdentifier(meta, "function");
+    this.next(); // eat `function`
 
-    if (this.prodParam.hasYield && this.eat(tt.dot)) {
+    if (this.prodParam.hasYield && this.match(tt.dot)) {
+      const meta = this.createIdentifier(
```

Comment:
Avoid creating an extra `meta` node when `tt.dot` is not met, instead we create `meta` when we have to. The same technique is also applied for `parseNewOrNewTarget`.